switchlink/switchlink_neigh_test.cc

```diff
@@ -0,0 +1,517 @@
+  }
+  if (intf_h) {
+    temp.intf_h = intf_h;
+  }
+  temp.opType = DELETE_NEIGHBOR;
+  temp.num_handler_calls++;
+  results.push_back(temp);
+}
+
+/*
+ * Dummy function for switchlink_db_get_interface_info(). This function
+ * is invoked by switchlink_process_neigh_msg() for getting the intf
+ * info from the database.
+ *
+ * Since this is a dummy method, we are passing an ifindex 1 to get L3
+ * interface info and ifindex 2 to get L2 interface info from the db.
+ *
+ * The actual function can also return SWITCHLINK_DB_STATUS_ITEM_NOT_FOUND
+ * in case it is not able to find the interface in the database.
+ * That scenario is mocked by passing an ifindex of 3.
+ */
+switchlink_db_status_t
+switchlink_db_get_interface_info(uint32_t ifindex,
+                                 switchlink_db_interface_info_t *intf_info) {
+  if (ifindex == 1) {
+    intf_info->intf_h = TEST_INTF_H;
+    intf_info->intf_type = SWITCHLINK_INTF_TYPE_L3;
+  } else if (ifindex == 2) {
+    intf_info->intf_h = TEST_INTF_H;
+    intf_info->bridge_h = TEST_BRIDGE_H;
+    intf_info->intf_type = SWITCHLINK_INTF_TYPE_L2_ACCESS;
+  } else if (ifindex == 3) {
+    intf_info = nullptr;
+    return SWITCHLINK_DB_STATUS_ITEM_NOT_FOUND;
+  }
+  return SWITCHLINK_DB_STATUS_SUCCESS;
+}
+
+/*
+ * Test fixture.
+ */
+class SwitchlinkNeighborTest : public ::testing::Test {
+protected:
+  struct nl_msg *nlmsg_ = nullptr;
+
+  // Sets up the test fixture.
+  void SetUp() override { ResetVariables(); }
+
+  // Tears down the test fixture.
+  void TearDown() override {
+    if (nlmsg_) {
+      nlmsg_free(nlmsg_);
+      nlmsg_ = nullptr;
+    }
+  }
+
+  void ResetVariables() {
+    // result variables
+    results.clear();
+  }
+};
+
+/*
+ * Creates an IPv4 neighbor
+ *
+ * Validates switchlink_process_neigh_msg(). It parses an
+ * RTM_NEWNEIGH message, invokes switchlink_create_neigh()
+ * for creating neighbor, nexthop and route entry.
+ *
+ * In create call...
+ * We invoke switchlink_create_mac()...
+ * ...and switchlink_create_neigh()....
+ * Hence we expect 2 test_results here, and this is the
+ * reason why test_results has been taken as a vector of size 2.
+ */
+TEST_F(SwitchlinkNeighborTest, createIPv4Neighbor) {
+  struct ndmsg hdr = {
+      .ndm_family = AF_INET,
+      .ndm_ifindex = 1,
+      .ndm_state = NUD_REACHABLE,
+      .ndm_flags = 0x1,
+      .ndm_type = 1,
+  };
+
+  const uint32_t ipv4_addr = IPV4_ADDR(10, 10, 10, 1);
+  const unsigned char mac_addr[6] = {0x00, 0xdd, 0xee, 0xaa, 0xdd, 0x00};
+
+  // Arrange
+  nlmsg_ = nlmsg_alloc_size(1024);
+  ASSERT_NE(nlmsg_, nullptr);
+  nlmsg_put(nlmsg_, 0, 0, RTM_NEWNEIGH, 0, 0);
+  nlmsg_append(nlmsg_, &hdr, sizeof(hdr), NLMSG_ALIGNTO);
+  nla_put_u32(nlmsg_, NDA_DST, htonl(ipv4_addr));
+  nla_put(nlmsg_, NDA_LLADDR, sizeof(mac_addr), &mac_addr);
+
+  // Act
+  const struct nlmsghdr *nlmsg = nlmsg_hdr(nlmsg_);
+  switchlink_process_neigh_msg(nlmsg, nlmsg->nlmsg_type);
+
+  // Assert
+  EXPECT_EQ(results.size(), 2);
```

Comment:
If `results.size()` is less than 2, you'll get out-of-bounds array accesses in the following checks.

I'm wondering whether the vector size test shouldn't be `ASSERT_EQ()` instead of `EXPECT_EQ()`.

Alternatively, you could do an `EXPECT_EQ(2)` here, `ASSERT_GE(1)` before you check `results[0]`, and `ASSERT_GE(2)` before you check `results[1]`.

Or maybe just abandon the current test case with `if (results.size() < N) return;`. I don't remember whether a failed assertion aborts the test case, the fixture, or the whole program. You'll have the failure of `EXPECT_EQ(2)` to tell you what happened.

Reply:
ASSERT_EQ() seems more befitting than EXPECT_EQ() to check results.size(). Thanks!